### DISCUSSION TITLE
Macro Support

### DISCUSSION THREAD

**Discussion by andymac4182** [markdown-confluence/markdown-confluence#297] [Ideas]
Hey,

I am looking at adding templates for common macros. Which macros do people want?

**Comment by dljsjr** [comment]
Obviously everyone's experiences might vary but these are my "top" macros based on frequency in my existing Confluence corpus:

1. Jira Issues (`jira`) macro; I'm *constantly* making pseudo-dashboards by embedding JIRA issues.
2. `expand` for collapsible sections that aren't callouts
3. TOC
4. Child Pages (this is actually 3 for me but reviewing our Confluence instance it seems like more people prefer TOC)
5. Layouts
6. Decision
7. Agile Wallboard Gadget

Stretch goals:
1. Integrate 1 with [this Jira integration](https://github.com/marc0l92/obsidian-jira-issue) (it provides an API in the global namespace already),
2. Integrate with [the updated Folder Notes fork](https://github.com/aidenlx/alx-folder-note) for 3 or 4
3. Integrate with something like [Multi-Column Markdown](https://github.com/ckRobinson/multi-column-markdown) or [Obsidian Columns](https://github.com/tnichols217/obsidian-columns) for 5 (in my experience Obsidian Columns seems to be more well behaved than multi-column markdown but multi-column has more installs, idk)

**Reply by andymac4182** [reply]
Expand support is already available https://markdown-confluence.com/features/callouts.html

Child pages is auto applied to any folder without a folder note. Is there anything else you are looking for?

I will have a look at the rest to see how much effort is involved in setting up the templates :) 

Thanks heaps for this feedback.

**Reply by dljsjr** [reply]
> Expand support is already available [markdown-confluence.com/features/callouts.html](https://markdown-confluence.com/features/callouts.html)

It's *really* ergonomically janky including any kind of complicated content inside of a callout in Obsidian. I frequently wrap `expand`s around other macros.

> Child pages is auto applied to any folder without a folder note. Is there anything else you are looking for?

Child pages for folders *with* a folder note :)

**Reply by andymac4182** [reply]
Do you have some examples of the sorts of pages you are publishing from Obsidian? It sounds like the more complex side so want to make sure it works nicely.

**Reply by dljsjr** [reply]
I'll see if I can produce an example that's both coherent and redacted 😅

**Reply by andymac4182** [reply]
Thanks @dljsjr It sounds like you have a fairly complex set of pages at the moment and I want to make sure the experience is good for those pages as well as the simpler ones.

**Reply by dljsjr** [reply]
Here's an example of a single block of ADF that uses a gamut of features to achieve the formatting I want:

<pre>
```adf
{
    "type": "expand",
    "attrs": {
        "title": "git diff"
    },
    "marks": [{
        "type": "breakout",
        "attrs": {
            "mode": "full-width"
        }
    }],
    "content": [{
        "type": "codeBlock", 
        "attrs": { "language": "diff", "layout": "full-width" },
        "content": [{
            "type": "text",
            "text": "&lt;redacted&gt;" 
        }]
    }]
}
```
</pre>

I use Jira issues w/ inline JQL a *lot*:

<pre>
```adf
{
  "type": "extension",
  "attrs": {
    "layout": "full-width",
    "extensionType": "com.atlassian.confluence.macro.core",
    "extensionKey": "jira",
    "parameters": {
      "macroParams": {
        "server": {
          "value": "System JIRA"
        },
        "columns": {
          "value": "key,summary,type,Status,resolution"
        },
        "columnIds": {
          "value": "issuekey,summary,issuetype,status,resolution"
        },
        "maximumIssues": {
          "value": "10"
        },
        "jqlQuery": {
          "value": "assignee=currentUser() AND resolution is empty"
        },
        "serverId": {
          "value": "&lt;redacted&gt;"
        }
      }
    }
  }
}
```
</pre>

Here's an example of where I have a folder that I would like to contain both a folder note and a listing of child pages:

<pre>
### Child Pages
```adf
    {
      "type": "extension",
      "attrs": {
        "layout": "default",
        "extensionType": "com.atlassian.confluence.macro.core",
        "extensionKey": "children",
        "parameters": {
          "macroParams": {
            "all": {
              "value": "true"
            },
            "style": {
              "value": "h4"
            },
            "sort": {
              "value": "creation"
            },
            "excerptType": {
              "value": "simple"
            }
          }
        }
      }
    }
```
## Tickets
```adf
{
  "type": "extension",
  "attrs": {
    "layout": "wide",
    "extensionType": "com.atlassian.confluence.macro.core",
    "extensionKey": "jira",
    "parameters": {
      "macroParams": {
        "server": {
          "value": "System JIRA"
        },
        "columns": {
          "value": "key,summary,type,Status,resolution"
        },
        "columnIds": {
          "value": "issuekey,summary,issuetype,status,resolution"
        },
        "maximumIssues": {
          "value": "20"
        },
        "jqlQuery": {
          "value": "key in (KEY-1, KEY-2, KEY-3)  "
        },
        "serverId": {
          "value": "&lt;redacted&gt;"
        }
      }
    }
  }
}
```
</pre>

![image](https://github.com/markdown-confluence/markdown-confluence/assets/444635/80c4ebab-fd67-473a-bca9-215bb6cb9f1b)

In most of my writing I try to keep individual documents really short and leverage hiearchy + linking to correlate things (maybe slightly inspired by Zettels/Atomic Notes?). Here, I wrote a design doc/proposal, then as I started the implementation, I kept notes on the WIP as a child page instead of making the existing page massive in size. I've found that most people don't wanna read long pages but are happy to click around in to a bunch of small pages so it just kinda works for me if I want to emphasize communication.

Hope that helps!

**Reply by andymac4182** [reply]
Thanks :) that helps massively. 

Honestly not sure the solution at this point but I will try set up this myself now so I can play with it and come up with something.
